### ISSUE TITLE
Dynamic symbols resolution via &DYNAMIC is potentially a problem in user code

### ISSUE THREAD

**Issue by vrurg** [perl6/problem-solving#127]
Dynamic symbols are currently being resolved via `&DYNAMIC` sub which is being looked up lexically. Due to its simple name it is not very unrealistic scenario where user code would define own sub of the same name resulting in "mysterious" failures and, eventually, complaints.

**Comment by jnthn** [comment]
We already have a mechanism which forces resolution of symbols from the setting to avoid accidental user overrides of core things; I think we can apply the same solution here when we code-gen the lookup a dynamic variable lookup.

**Comment by vrurg** [comment]
I would suggest moving the sub into PseudoStash and making it a static method. Correspondingly, `Actions` code would produce method calls for dynamic lookups.

**Comment by vrurg** [comment]
@jnthn you're fast! 

Actually, Actions does this:

```
            $past := QAST::Op.new(
                :op('call'), :name('&DYNAMIC'),
                $*W.add_string_constant($name));
```

Do you suggest replacing `:name('&DYNAMIC')` with `$*W.find_symbol(:setting-only)`? I'm simplifying because don't remember if the op would work this way.

Anyway, I like the idea of moving it into `PseudoStash` a bit more because they basically do the same thing. Actually, I'm currently trying to unify `PseudoStash` dynamic lookups with `&DYNAMIC`.

**Comment by jnthn** [comment]
> Do you suggest replacing :name('&DYNAMIC') with $*W.find_symbol(:setting-only)? I'm simplifying because don't remember if the op would work this way.

Yes; the `:name('&DYNAMIC')` would go away, and the first child be a `QAST::WVal` whose value is the result of the `find_setting` lookup. Only thing we may run into is bootstrapping shenanigans, since this depends on `&DYNAMIC` having been defined before the first use of a dynamic variable.

> Anyway, I like the idea of moving it into PseudoStash a bit more because they basically do the same thing.

It'd have worse code size, though.

**Comment by vrurg** [comment]
> > Anyway, I like the idea of moving it into PseudoStash a bit more because they basically do the same thing.
> 
> It'd have worse code size, though.

This is my only worry about the approach. And possible performance hit.

Actually, I didn't mention another option. `&DYNAMIC` could be renamed to something like `&RESOLVE-DYN-SYMS` which due to its terrible name would reduce the chances of accidental redefinition. In a way, I even like it from the point of view of it allowing for some nice tricks. ;)

**Comment by jnthn** [comment]
> This is my only worry about the approach.

It's worth worrying about, since it can have fallout.

> And possible performance hit.

True, a method lookup would be a bit more costly.

> Actually, I didn't mention another option. &DYNAMIC could be renamed to something like &RESOLVE-DYN-SYMS which due to its terrible name would reduce the chances of accidental redefinition. In a way, I even like it from the point of view of it allowing for some nice tricks. ;)

If we're going to explicitly allow it, we may as well keep it called `&DYNAMIC`. I mean, it's already well established that uppercase names are reserved.
